documentation/api/admin/v1/delete.markdown

```diff
@@ -0,0 +1,54 @@
+---
+title: "Delete endpoint"
+layout: default
+---
+
+[curl]: ../../query/curl.html#using-curl-from-localhost-non-sslhttp
+
+The `/delete` endpoint can be used to trigger the immediate deletion of all data
+associated with a certname. It is important to note that the delete operation
+doesn't account for in flight commands which may be in the command queue but not
+yet processed by PuppetDB, the targeted node will reappear when the in flight
+command is processed in this case.
+
+## `POST /pdb/admin/v1/delete`
+
+The delete operation must be triggered by a POST.
+
+### Request format
+
+The POST request should specify `Content-Type: application/json` and
+the request body should look like this:
+
+``` json
+{"version" : 1, "command" : "delete", "payload" : {"certname" : CERTNAME_TO_DELETE}}
+```
+
+Where `CERTNAME_TO_DELETE` is a string of the certname to be deleted.
+
+### URL parameters
+
+* The POST endpoint accepts no URL parameters.
+
+### Response format
+
+The response type will be `application/json`, and upon success will
+include this JSON map:
+
+``` json
+{"ok": true}
```

Comment:
One other concern I have about the wire format and this response, is both implied to me that its a command that gets enqueued on the command queue. I would then assume if I just stopped the puppet-agent daemon on that machine, and then sent this command, it would be processed after all the previous commands were processed from the last puppet run, but that might not be the case if I'm running any queue depth.

I'm just throwing things out wildly, but we could start with a simpler wire format
```
{"version": 1, "certname": <string>}
```
which would allow us to take up that more complicated wire format in v2 if we ever wanted to treat it as a command.

If we simplify the wire format and don't enqueue it on the command queue, it'd be nice to return a bit of information to them, ie. the JSON representation of "deleted 1 certname, dead-server.company.net in 747ms" since it won't be part of the command metrics.

Reply:
On the other side, instead of changing the wire format, we could do what the wire format implies and enqueue it on the command queue (this wouldn't be a command we need to sync we can just document that it is ephemeral). This style would make it easier for someone to write a Bolt plan to, first go out and disable the agent service on all the machines they want to delete and then enqueue the delete commands.

Reply:
The other admin /cmd endpoint uses a similar wire format so I was trying to stay consistent with that. I was thinking that having the :payload key would allow us to include keys for more targeted deletions (say just factsets/reports or a vector of certnames) later on, but you're right we could just rev the endpoint and add it when we decide we want/need it.  

I agree with the part about returning more detailed information about what we deleted in the response. I'll update that. 

Also an interesting idea about submitting the command to the queue, that would account for some of the edge cases if a command for the certname was in the queue already, etc. The ticket seems like it was asking for a way to immediately delete data without having to hit the db with a sql statement directly. I'm kinda leaning towards having it delete the thing directly, though we could ask around and see what the folks using it would prefer.

Reply:
Ah I see, that's a general endpoint `/pdb/admin/v1/cmd` for they need to specify the specific command `"expire_nodes", "purge_nodes", "purge_reports"` etc. Maybe this command belongs on that cmd endpoint then? I don't like the response from that one either, but it does provide the protection of
> Only one maintenance operation can be running at a time. Any request received while an operation is already in progress will return an HTTP conflict status (409).

If we leave it on this endpoint I'd suggest at least dropping the key `"command" : "delete"`  because the url `/admin/v1/delete` specifies the command is "delete". If we add other entities to delete in the future we'll have to rev the cmd version no matter what and can add it back in then.

Reply:
that's a good idea, I'll move it to the general `/cmd` endpoint

Reply:
Hmm, yeah, I hadn't really thought about it hard yet, but I think I'd probably been vaguely imagining it might just be a `/cmd` operation, but looking back at how `/cmd` now works, I suppose you could argue that might add complexity (including perhaps wrt careful error handling) without sufficient justification (i.e. if you can mix this in with the other commands).  

On the other hand if it's "sufficiently similar" it might be fine, and easily share the infrastructure.  It also crossed my mind that at some point "delete by query" might be a natural extension if/when we ever need it, and that would be more complicated.

I suppose a slight variation on what we have here would be a `admin/v1/delete-nodes` which would only need the certname(s) specification(s) in the body, and could also support (and I suppose could even just start with) one of our query-style shortcut urls, i.e. `admin/v1/delete-nodes/CERTNAME.  Then we wouldn't need any command body to start with.

...of course people might well suggest that this should be a an http `DELETE` on something like `.../node/CERTNAME`, and likely wouldn't be happy with a `GET` or `POST`.

Reply:
How important do y'all feel locking the `"delete"` command down to one request at a time is? 

I just realized we probably don't want to use the existing `clean-lock` that gates admin `"clean"` commands because the lock also blocks during normally scheduled gc operations which could block `"delete"` commands when we didn't intend to. 

If we do think it's important to limit the amount of concurrent `"delete"` commands we process we can gate it behind a lock of its own or a a semaphore if we want to let a few go at once, etc.